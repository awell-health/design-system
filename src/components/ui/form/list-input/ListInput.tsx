import { useEffect, useState } from 'react';
import { Icon } from '../../icon';
import { cn } from '../../../../lib/utils';
import { FC } from 'react';
interface Props {
  value?: string[];
  className?: string;
  pillClassName?: string;
  onChange?: (items: string[]) => void;
  placeholder?: string;
  helperText?: string;
  error?: string;
  required?: boolean;
  label?: string;
}

const ListInput: FC<Props> = ({
  value,
  pillClassName,
  className,
  onChange,
  error,
  required,
  label,
  placeholder = 'Type to add an item',
  helperText = 'Hit enter or click + to add a new item'
}) => {
  const [items, setItems] = useState<string[]>(value ?? []);
  const [inputValue, setInputValue] = useState<string>('');

  useEffect(() => {
    if (onChange) {
      onChange(items);
    }
  }, [items]);

  const addItem = () => {
    if (inputValue) {
      setItems([...items, inputValue]);
      setInputValue('');
    }
  };

  const removeItem = (index: number) => {
    const updatedItems = [...items].filter((_, i) => i !== index);
    setItems(updatedItems);
  };

  return (
    <div className='flex flex-col gap-1'>
      {label && (
        <div className='text-sm font-medium text-slate-700'>
          {label} {required && <span className='text-red-500'>*</span>}
        </div>
      )}
      <div
        className={cn(
          'relative flex gap-2 items-center',
          'shadow border border-slate-300 px-1 py-1 min-h-[38px] rounded-lg',
          error && 'border-red-500',
          className
        )}
      >
        <div className='flex gap-2 flex-wrap'>
          {items.map((item, index) => (
            <div
              key={index}
              className={cn(
                'bg-slate-100 rounded-lg flex gap-1 px-1 text-sm items-center',
                pillClassName
              )}
            >
              <span className='py-0.5 px-1.5'>{item}</span>
              <Icon
                icon='RiCloseFill'
                className='cursor-pointer rounded-full hover:bg-slate-300'
                size={16}
                onClick={() => removeItem(index)}
              />
            </div>
          ))}
        </div>
        <input
          className={cn(
            'text-sm font-normal',
            'border-none shadow-none',
            'focus:outline-none focus:ring-0 focus:ring-offset-0',
            'placeholder:text-xs'
          )}
          placeholder={placeholder}
          value={inputValue}
          onChange={(e) => setInputValue(e.target.value)}
          onKeyDown={(e) => {
            if (e.key === 'Enter') {
              addItem();
            }
          }}
        />
        <div className='ml-auto'>
          <Icon
            icon='RiAddFill'
            size={16}
            className='cursor-pointer rounded-full hover:bg-slate-300'
            onClick={addItem}
          />
        </div>
      </div>
      {helperText && <div className='text-xs text-slate-500'>{helperText}</div>}
      {error && <div className='text-xs text-red-500'>{error}</div>}
    </div>
  );
};

export { ListInput };
