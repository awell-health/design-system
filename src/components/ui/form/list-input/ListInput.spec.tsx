import { fireEvent, render } from '@testing-library/react';
import { expect, it, describe, vi } from 'vitest';
import { ListInput } from './ListInput';
import React from 'react';

describe('Textarea', () => {
  const subject = (props = {}) => {
    const defaultProps = {
      value: ['test'],
      onChange: () => {},
      placeholder: 'Type to add an item',
      helperText: 'Hit enter or click + to add a new item',
      label: 'List Input',
      required: true
    };

    return render(<ListInput {...defaultProps} {...props} />);
  };
  it('match snapshot', () => {
    expect(subject().container).toMatchSnapshot();
  });

  it('shows error state', () => {
    const { container, queryByText } = subject({ error: 'Error message' });

    expect(container.querySelectorAll('.border-red-500')).toHaveLength(1);
    expect(queryByText('Error message')).toBeInTheDocument();
  });

  it('returns the correct value', () => {
    const onChange = vi.fn();
    const { container, queryByText } = subject({ onChange });

    expect(queryByText('test')).toBeInTheDocument();

    const input = container.querySelector('input');

    if (!input) {
      throw new Error('Input not found');
    }

    fireEvent.change(input, { target: { value: 'test4' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(queryByText('test4')).toBeInTheDocument();
    expect(onChange).toHaveBeenCalledWith(['test', 'test4']);
  });
});
