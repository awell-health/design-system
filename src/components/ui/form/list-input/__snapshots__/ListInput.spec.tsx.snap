// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Textarea > match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-1"
  >
    <div
      class="text-sm font-medium text-slate-700"
    >
      List Input
       
      <span
        class="text-red-500"
      >
        *
      </span>
    </div>
    <div
      class="relative flex gap-2 items-center shadow border border-slate-300 px-1 py-1 min-h-[38px] rounded-lg"
    >
      <div
        class="flex gap-2 flex-wrap"
      >
        <div
          class="bg-slate-100 rounded-lg flex gap-1 px-1 text-sm items-center"
        >
          <span
            class="py-0.5 px-1.5"
          >
            test
          </span>
          <svg
            class="remixicon cursor-pointer rounded-full hover:bg-slate-300"
            fill="default"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.9997 10.5865L16.9495 5.63672L18.3637 7.05093L13.4139 12.0007L18.3637 16.9504L16.9495 18.3646L11.9997 13.4149L7.04996 18.3646L5.63574 16.9504L10.5855 12.0007L5.63574 7.05093L7.04996 5.63672L11.9997 10.5865Z"
            />
          </svg>
        </div>
      </div>
      <input
        class="text-sm font-normal border-none shadow-none focus:outline-none focus:ring-0 focus:ring-offset-0 placeholder:text-xs"
        placeholder="Type to add an item"
        value=""
      />
      <div
        class="ml-auto"
      >
        <svg
          class="remixicon cursor-pointer rounded-full hover:bg-slate-300"
          fill="default"
          height="16"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 11V5H13V11H19V13H13V19H11V13H5V11H11Z"
          />
        </svg>
      </div>
    </div>
    <div
      class="text-xs text-slate-500"
    >
      Hit enter or click + to add a new item
    </div>
  </div>
</div>
`;
